### PR TITLE
Jpe/update mag tests

### PR DESCRIPTION
**Pull request by jedman**
update `test_forward_Mag_Linear.py` to test different simulation engines and parallelism

### CODE REVIEW

**Comment by jedman** [comment]
@santisoler -- these tests pass for `geoana`, but I'm seeing failures (based on the numerical comparisons) for all the `choclo` versions

File: SimPEG/regularization/pgi.py [hunk 1]
```diff
@@ -94,21 +94,35 @@ class PGIsmallness(Smallness):
...
     extracts the GMM parameters for the most representative rock unit within each active cell
-    in the :class:`RegularizationMesh`. The regularization function (objective function) for
-    ``PGIsmallness`` is given by:
+    in the :class:`RegularizationMesh`.
+
+    When the ``approx_eval`` property is ``True``, we assume the physical property distributions of each geologic units
+    are distinct (no significant overlap of their respective physical properties distribution). The GMM probability
+    density value at any each point of the physical property space can then be approximated by the locally dominant
+    Gaussian distribution. In this case, the PGI regularization function (objective function) can be expressed as a
+    least-square:
```
Comment (santisoler) [inline SimPEG/regularization/pgi.py]: I think these lines belong to another PR. Can you revert them back to the original state since they aren't related to the mag simulation?

File: SimPEG/regularization/pgi.py [hunk 2]
```diff
@@ -94,21 +94,35 @@ class PGIsmallness(Smallness):
...
     .. math::
-        \phi (\mathbf{m}) = \frac{1}{2}
-        \big [ \mathbf{m} - \mathbf{m_{ref}}(\Theta, \mathbf{z}^\ast ) \big ]^T
-        \mathbf{W} ( \Theta , \mathbf{z}^\ast ) \,
-        \big [ \mathbf{m} - \mathbf{m_{ref}}(\Theta, \mathbf{z}^\ast ) \big ]
+        \phi (\mathbf{m}) &= \frac{\alpha_{pgi}}{2}
+        \big | \mathbf{W} ( \Theta , \mathbf{z}^\ast ) \, (\mathbf{m} - \mathbf{m_{ref}}(\Theta, \mathbf{z}^\ast ) \, \Big \|^2
+        &+ \sum_{j=x,y,z} \frac{\alpha_j}{2} \Big \| \mathbf{W_j G_j \, m} \, \Big \|^2 \\
+        &+ \sum_{j=x,y,z} \frac{\alpha_{jj}}{2} \Big \| \mathbf{W_{jj} L_j \, m} \, \Big \|^2
+        \;\;\;\;\;\;\;\; \big ( \textrm{optional} \big )
```
Comment (santisoler) [inline SimPEG/regularization/pgi.py]: Same for these lines

File: SimPEG/regularization/pgi.py [hunk 3]
```diff
@@ -94,21 +94,35 @@ class PGIsmallness(Smallness):
...
         - :math:`\mathbf{W}(\Theta , \mathbf{z}^\ast )` is a weighting matrix.
 
+    For the full, non-approximated PGI regularization, please refer to
+    (`Astic, et al 2019 <https://owncloud.eoas.ubc.ca/s/TMB3Jdr8ScqSPm7/download>`__;
+    `Astic et al 2020 <https://owncloud.eoas.ubc.ca/s/PAxpHQt7CGk6zT4/download>`__).
+
+    When the ``approx_eval`` property is ``True``, you may also set the ``approx_gradient`` and ``approx_hessian``
+    properties to ``True`` so that the least-squares approximation is used to compute the gradient, as it is making the
+    same assumptions about the GMM.
+
```
Comment (santisoler) [inline SimPEG/regularization/pgi.py]: Same for these lines

File: SimPEG/regularization/pgi.py [hunk 4]
```diff
@@ -118,17 +132,6 @@ class PGIsmallness(Smallness):
...
-    values that are uncorrelated. In this case, the weighting matrix is diagonal and the
-    regularization function (objective function) can be expressed as:
-
-    .. math::
-        \phi (\mathbf{m}) = \frac{1}{2} \Big \| \mathbf{W}_{\! 1/2}(\Theta, \mathbf{z}^\ast ) \,
-        \big [ \mathbf{m} - \mathbf{m_{ref}}(\Theta, \mathbf{z}^\ast ) \big ] \, \Big \|^2
-
-    When the ``approx_eval`` property is ``True``, you may also set the ``approx_gradient`` property
-    to ``True`` so that the least-squares approximation is used to compute the gradient.
-
```
Comment (santisoler) [inline SimPEG/regularization/pgi.py]: Same for these lines

File: SimPEG/regularization/pgi.py [hunk 5]
```diff
@@ -142,19 +145,8 @@ class PGIsmallness(Smallness):
...
-    covariances are zero and we can use a weighting matrix of the form:
-
-    .. math::
-        \mathbf{W}_{\! 1/2}(\Theta ,{\mathbf{z}^\ast } ) = diag \Big ( \big [ \mathbf{v \odot w}
-        \odot \boldsymbol{\sigma}_{\mathbf{z}^\ast}^{-2} \big ]^{1/2} \Big )
-
-    where :math:`\boldsymbol{\sigma}_{\mathbf{z}^\ast}^2` are the variances extracted using the
-    membership array :math:`\mathbf{z}^\ast`.
+        \mathbf{W}(\Theta ,{\mathbf{z}^\ast } ) = \boldsymbol{\Sigma}_{\mathbf{z^\ast}}^{\frac{-1}{2}} \,
+        diag \big ( \mathbf{w} \big )
```
Comment (santisoler) [inline SimPEG/regularization/pgi.py]: Same for these lines

File: tests/pf/test_forward_Mag_Linear.py [hunk 6]
```diff
@@ -1,77 +1,162 @@
...
+    return H0, b0
+
+
+def get_block_inds(grid: np.ndarray, block: np.ndarray) -> np.ndarray:
+    """
+    get the indices for a block
+
+    :param grid: (n, 3) array of xyz locations
+    :param block: (3, 2) array of (xmin, xmax), (ymin, ymax), (zmin, zmax) dimensions of the block
+    :return boolean array of
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]: Docstrings in these utils functions and fixtures are much appreciated! In SimPEG we use [numpydoc style](https://numpydoc.readthedocs.io/en/latest/format.html). You might see some other styles in SimPEG's docstrings, but we are enforcing numpydocs in new PRs.

Would you mind changing them to numpydocs style? I think it can be automated with tools like [`pydocstring`](https://github.com/robodair/pydocstring)

**Review by santisoler — CHANGES_REQUESTED** [review]
Hi @jedman! Thanks for writing all these tests! They helped me to find two bugs in the code: 
1. In the forward function I was overwritting the elements of `field` instead of incrementing their value with the current result for the current cell.
2. Also in the forward function, the `index_offset` had a wrong indentation, so the results weren't being written in the right position in the `fields` array.

I've pushed fixes to these bugs to my `mag-choclo` branch. Feel free to update this branch and run the tests again.

The only two tests that are failing are related to the non-implemented gradiometry component.
Since I don't know if I'm going to add those in my PR (I need to add their corresponding kernels to Choclo first), maybe it would be better to avoid those tests, but check if they raise a `NotImplementedError`.

As I see, we are not testing the building of the sensitivity matrix. It would be nice to add those tests. We can just parametrize the tests functions to run with `store_sensitivities="ram"` and `store_sensitivities="forward_only"` (see https://github.com/simpeg/simpeg/blob/d4612044332d6e5e56f193c804e38240cf2ff329/tests/pf/test_forward_Grav_Linear.py#L111 as an example).

I leave some minor suggestions down here.

Thanks again for this!

File: SimPEG/regularization/pgi.py (hunk 1 above)
Comment (jedman) [reply SimPEG/regularization/pgi.py]: The commits with these changes (and the lines below) are merged into SimPEG `main` -- if you sync your fork the diff should go away

File: tests/pf/test_forward_Mag_Linear.py (hunk 6 above)
Comment (jedman) [reply tests/pf/test_forward_Mag_Linear.py]: sure, no problem!

**Comment by jedman** [comment]
Thanks @santisoler ! I'll add tests for the sensitivities, reformat the docstrings, and omit the gradiometry tests for `choclo`

**Comment by jedman** [comment]
@santisoler , I added tests for the different ways of storing sensitivities, and some of them fail with small numerical discrepancies (differences ~1e-6). I think it makes sense to relax the tolerances a little, but let me know what you think

File: tests/pf/test_forward_Mag_Linear.py [hunk 7]
```diff
@@ -85,14 +228,15 @@ def get_block_inds(grid, block):
     np.testing.assert_allclose(d_t, d_t2)  # double check internal projection
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py:228]: Move this line to the end
```suggestion
```

Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]: Move this line to the bottom
```suggestion
```

File: tests/pf/test_forward_Mag_Linear.py [hunk 8]
```diff
@@ -85,14 +228,15 @@ def get_block_inds(grid, block):
...
     d = (
-        prism_1.magnetic_flux_density(locXyz)
-        + prism_2.magnetic_flux_density(locXyz)
-        + prism_3.magnetic_flux_density(locXyz)
+        prism_1.magnetic_flux_density(receiver_locations)
+        + prism_2.magnetic_flux_density(receiver_locations)
+        + prism_3.magnetic_flux_density(receiver_locations)
     )
 
     np.testing.assert_allclose(d_x, d[:, 0])
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]: Replace these lines for:
```python
    # Check results
    rtol, atol = 1e-7, 1e-6
    np.testing.assert_allclose(
        d_t, d_t2, rtol=rtol, atol=atol
    )  # double check internal projection
    np.testing.assert_allclose(d_x, d[:, 0], rtol=rtol, atol=atol)
    np.testing.assert_allclose(d_y, d[:, 1], rtol=rtol, atol=atol)
    np.testing.assert_allclose(d_z, d[:, 2], rtol=rtol, atol=atol)
    np.testing.assert_allclose(d_t, d @ tmi, rtol=rtol, atol=atol)
```

File: tests/pf/test_forward_Mag_Linear.py [hunk 9]
```diff
@@ -101,177 +245,144 @@ def get_block_inds(grid, block):
...
+            + prism_2.magnetic_field_gradient(receiver_locations)
+            + prism_3.magnetic_field_gradient(receiver_locations)
+        ) * mu_0
+
+        np.testing.assert_allclose(d_xx, d[..., 0, 0], rtol=1e-10, atol=1e-12)
+        np.testing.assert_allclose(d_xy, d[..., 0, 1], rtol=1e-10, atol=1e-12)
+        np.testing.assert_allclose(d_xz, d[..., 0, 2], rtol=1e-10, atol=1e-12)
+        np.testing.assert_allclose(d_yy, d[..., 1, 1], rtol=1e-10, atol=1e-12)
+        np.testing.assert_allclose(d_yz, d[..., 1, 2], rtol=1e-10, atol=1e-12)
+        np.testing.assert_allclose(d_zz, d[..., 2, 2], rtol=1e-10, atol=1e-12)
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]:
```suggestion
        # Check results
        rtol, atol = 1e-7, 1e-6
        np.testing.assert_allclose(d_xx, d[..., 0, 0], rtol=rtol, atol=atol)
        np.testing.assert_allclose(d_xy, d[..., 0, 1], rtol=rtol, atol=atol)
        np.testing.assert_allclose(d_xz, d[..., 0, 2], rtol=rtol, atol=atol)
        np.testing.assert_allclose(d_yy, d[..., 1, 1], rtol=rtol, atol=atol)
        np.testing.assert_allclose(d_yz, d[..., 1, 2], rtol=rtol, atol=atol)
        np.testing.assert_allclose(d_zz, d[..., 2, 2], rtol=rtol, atol=atol)
```

File: tests/pf/test_forward_Mag_Linear.py [hunk 10]
```diff
@@ -281,89 +392,68 @@ def get_block_inds(grid, block):
     np.testing.assert_allclose(data[:, 3], d @ tmi)
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]: Replace these lines with:
```python
    # Check results
    rtol, atol = 9e-6, 3e-7
    np.testing.assert_allclose(data[:, 0], d[:, 0], rtol=rtol, atol=atol)
    np.testing.assert_allclose(data[:, 1], d[:, 1], rtol=rtol, atol=atol)
    np.testing.assert_allclose(data[:, 2], d[:, 2], rtol=rtol, atol=atol)
    np.testing.assert_allclose(data[:, 3], d @ tmi, rtol=rtol, atol=atol)
```

File: tests/pf/test_forward_Mag_Linear.py [hunk 11]
```diff
@@ -281,89 +392,68 @@ def get_block_inds(grid, block):
...
-        prism_1.magnetic_flux_density(locXyz)
-        + prism_2.magnetic_flux_density(locXyz)
-        + prism_3.magnetic_flux_density(locXyz)
+        prism_1.magnetic_flux_density(receiver_locations)
+        + prism_2.magnetic_flux_density(receiver_locations)
+        + prism_3.magnetic_flux_density(receiver_locations)
     )
     d_amp = np.linalg.norm(d, axis=1)
 
     np.testing.assert_allclose(data, d_amp)
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py]:
```suggestion
    # Check results
    rtol, atol = 1e-7, 1e-6
    np.testing.assert_allclose(data, d_amp, rtol=rtol, atol=atol)
```

File: tests/pf/test_forward_Mag_Linear.py [hunk 12]
```diff
@@ -1,77 +1,220 @@
-import unittest
 import discretize
```
Comment (santisoler) [inline tests/pf/test_forward_Mag_Linear.py:3]: We should use these imports and update the type hints so they can work on Python 3.8
```suggestion
from typing import List, Tuple
import discretize
```

**Review by santisoler — CHANGES_REQUESTED** [review]
Thanks @jedman for these changes. I reproduced the same differences in the tests. You're right, we should relax a few tolerances. Most of the failures are related to comparison between values ~ 0.  A way to overcome those is to set an `atol` different than zero, to something appropriate to the values we are testing against. I've added some suggestions for `rtol` and `atol` values, find them below. Let me know if any of those fail on your side.

Another thing we should change is the type hints. Since SimPEG supports Python >= 3.8, we need to import `List` and `Tuple` from `typing` and use those in the hints.

I've updated my `mag-choclo` branch against the latest changes in `simpeg/main`, so feel free to update this one too.

After these changes, I think we should merge this. We can keep working on tests later on if we need to. Thanks again for the help!

**Comment by jedman** [comment]
Thanks @santisoler , I made the changes you suggested, and everything is passing for me now. Please let me know if there are any other ways I can contribute!

**Review by santisoler — APPROVED** [review]

**Comment by santisoler** [comment]
Thanks a lot @jedman for the help! I'm merging this PR.

I'll keep working on the mag simulation, specially after we merge https://github.com/simpeg/simpeg/pull/1285. I'll ping you asking for future reviews, and maybe asking some code too. Thanks again for the help and contributions!
